crypto/x509/x509_vfy.c

```diff
@@ -1752,18 +1753,23 @@ static int internal_verify(X509_STORE_CTX *ctx)
      * is allowed to reset errors (at its own peril).
      */
     while (n >= 0) {
-        EVP_PKEY *pkey;
-
         /*
-         * Skip signature check for self signed certificates unless explicitly
-         * asked for.  It doesn't add any security and just wastes time.  If
-         * the issuer's public key is unusable, report the issuer certificate
+         * Skip signature check for self-signed certificates unless explicitly
+         * asked for because it does not add any security and just wastes time.
+         * If the issuer's public key is not available or its key usage does
+         * not support issuing the subject cert, report the issuer certificate
          * and its depth (rather than the depth of the subject).
          */
         if (xs != xi || (ctx->param->flags & X509_V_FLAG_CHECK_SS_SIGNATURE)) {
+            EVP_PKEY *pkey;
+            int issuer_depth = n + (xi == xs ? 0 : 1);
+            int ret = x509_signing_allowed(xi, xs);
```

Comment:
Perhaps the signing-allowed check should happen only at `issuer_depth > 0`?  Otherwise an EE-self-signed cert might fail when `X509_V_FLAG_CHECK_SS_SIGNATURE` is set, and the keyUsage does not include `KU_KEY_CERT_SIGN`.  And I think the goal of this PR is allow such EE certs.  I don't think that `X509_V_FLAG_CHECK_SS_SIGNATURE` should override that policy.

Reply:
Good thought not to check key usage restrictions for root certs regardless of `X509_V_FLAG_CHECK_SS_SIGNATURE`.
I've done this tentatively, and all tests still pass.

Before suggesting to forward-port ;-) this to the master, I plan to check if this is RFC 5280 compliant...

Reply:
Thanks!  Indeed a forward-port to master may be appropriate.  With this much more focused PR, it was easier to separate heat from chaff and notice this potential anomaly.  I'll be happy to review your findings.  Good luck.

Reply:
Note that my comment wasn't broadly about self-signatures on root CA, but just about the depth 0 case.  But presumably if the CA is not at depth 0 then the key usage would be checked anyway to allow it to sign the immediate subject certificate, right?  If so, it may be appropriate to just skip the check for the top of chain self-signed TA regardless of the chain depth.

Reply:
> Note that my comment wasn't broadly about self-signatures on root CA, but just about the depth 0 case. 

Sure.

> But presumably if the CA is not at depth 0 then the key usage would be checked anyway to allow it to sign the immediate subject certificate, right?

Right. I've just verified with RFC 5280 that checking the keyUsage keyCertSign is demanded during chain validation for a cert of depth n in preparation of checking cert n+1.
Quoting from https://tools.ietf.org/html/rfc5280#section-6.1.4:
```
(n) If a key usage extension is present, verify that the
              keyCertSign bit is set.
```

Moreover, I understand this in a way that this step is not meant to be done if there is no such cert n+1, i.e., in our special case of a single self-signed (or at least self-issued) EE cert.
So to address my above concern of RFC 5280 compliance, I've slightly adapted the fixup to do the key usage check also for depth 0 certs in case there are further certs in the given chain.
The tests still pass on the adapted version.

Reply:
> If so, it may be appropriate to just skip the check for the top of chain self-signed TA regardless of the chain depth.

I did not get what you mean here?

Reply:
Here is also an interesting further requirement of RFC 5280 regarding verifying cert signatures and key usage checks:
```
If the keyUsage extension is present, then the subject public key
   MUST NOT be used to verify signatures on certificates or CRLs unless
   the corresponding keyCertSign or cRLSign bit is set
```

Reply:
On the other hand from the perspective of RFC5280, a trust anchor is just a public key, trust anchors in the form of self-signed certificates are somewhat outside the model. Verifying self-signatures is therefore in good measure also outside the model.  I'll take a closer look at the updated code soon.

Reply:
I wrote yesterday:

> Here is also an interesting further requirement of RFC 5280 regarding verifying cert signatures and key usage checks:
> 
> ```
> If the keyUsage extension is present, then the subject public key
>    MUST NOT be used to verify signatures on certificates or CRLs unless
>    the corresponding keyCertSign or cRLSign bit is set
> ```

On this it looks like we have a problem that I'm currently investigating:
Adding such a tests makes a couple of tests fail.
Therefore in the tentative commit that I did pushed I've commented out the new check `!x509_signing_allowed(xi, xs)`.

Reply:
It turns out that the mentioned problem with some tests failing was just due to a bug interpreting the result of `x509_signing_allowed()`.

Reply:
> Perhaps the signing-allowed check should happen only at `issuer_depth > 0`? Otherwise an EE-self-signed cert might fail when `X509_V_FLAG_CHECK_SS_SIGNATURE` is set, and the keyUsage does not include `KU_KEY_CERT_SIGN`. And I think the goal of this PR is allow such EE certs. I don't think that `X509_V_FLAG_CHECK_SS_SIGNATURE` should override that policy. Am I missing something?

It turns out that we must not skip the signing-allowed check because of the mentioned extra requirement in https://tools.ietf.org/html/rfc5280#section-4.2.1.3:
```
If the keyUsage extension is present, then the subject public key
   MUST NOT be used to verify signatures on certificates or CRLs unless
   the corresponding keyCertSign or cRLSign bit is set
```
So we have to obey key usage restrictions.
In case of a self-issued root/EE cert at the end of the chain, what we can do in case `X509_V_FLAG_CHECK_SS_SIGNATURE` is to simply skip verifying the (self-)signatures of such a certs. This is what I've now implemented. 

Moreover I've re-phrased the existing checks and added some comments for better readability of `internal_verify()`.

Reply:
As planned above in https://github.com/openssl/openssl/pull/12357#discussion_r449687809 I've meanwhile prepared a 'forward-port' of this last commit in #12375.

@romen, note that I've followed your practice and use as the number of that related PR the value that results from swapping the last two digits :wink: